[Security Solution] add new route for fetching results by index pattern

kapral18 opened the pull request:
Address #183698

This MR introduces a new internal DQD `GET` endpoint for fetching index results by pattern without aggregation in descending order by `@timestamp`:

`GET  /internal/ecs_data_quality_dashboard/results/:pattern`

This new endpoint supports additional query params like:

 - `from` - es search pagination offset
 - `size`- es search query response size
 - `outcome` - es search query response filter by presence/absence of `incompatibleFieldCount` field
 - `startDate` - es search query response filter by `@timestamp` (inclusive)
 - `endDate`- es search query response filter by `@timestamp` (inclusive)
 
This endpoint returns a list of historical index `results` sorted by `@timestamp` and a `total` number of results for facilitating pagination on client.
 
Additionally this MR updates latest results endpoint to have simplified naming for improved consistency, lifts shared functionality into a helper function and adds necessary unit tests.  

In particular:

- `GET  /internal/ecs_data_quality_dashboard/results/indices_latest/:pattern` has been simplified to just `GET  /internal/ecs_data_quality_dashboard/results_latest/:pattern`. All respective server and client code has been updated.
  
## How to test

Example:
 - `GET  /internal/ecs_data_quality_dashboard/results/logs-*?from=5&size=5&outcome=fail&startDate=now-1w/w&endDate=now`
 - `GET  /internal/ecs_data_quality_dashboard/results_latest/logs-*`

kapral18 committed:
[Security Solution] add new route for fetching results by index pattern

Address #183698

- Introduce a new constant `RESULTS_INDICES_PATTERN_ROUTE_PATH`.
- Add a helper function `getAuthorizedIndexNames` to fetch authorized
index names.
- Update `getResultsIndicesLatestRoute` to use the new helper function.
- Create a new route `getResultsIndicesPatternRoute` for fetching
results by index pattern.
- Add tests for the new route.
- Update schemas to include new parameters and query types for the new
route.

elasticmachine commented:
Pinging @elastic/security-threat-hunting (Team:Threat Hunting)

semd commented on x-pack/plugins/ecs_data_quality_dashboard/server/schemas/result.ts:
```diff
@@ -39,5 +39,17 @@ export type ResultDocument = t.TypeOf<typeof ResultDocument>;
+  pattern: t.string,
+});
+
+export const GetResultsIndicesPatternQuery = t.type({
+  limit: t.union([t.number, t.undefined]),
+  from: t.union([t.number, t.undefined]),
+  startDate: t.union([t.string, t.undefined]),
+  endDate: t.union([t.string, t.undefined]),
+  outcome: t.union([t.literal('pass'), t.literal('fail'), t.undefined]),
+});
```
NIT: We can use `t.partial` so we don't have to do the union with `undefined` in each parameter:

```suggestion
export const GetResultsIndicesPatternQuery = t.partial({
  limit: t.number,
  from: t.number,
  startDate: t.string,
  endDate: t.string,
  outcome: t.union([t.literal('pass'), t.literal('fail')),
});
```

kapral18 replied:
thanks for catching, will change

semd commented on x-pack/plugins/ecs_data_quality_dashboard/server/routes/results/get_results_indices_pattern.ts:
```diff
@@ -0,0 +1,182 @@
+type NonNullableQueryBool = NonNullable<NonNullableQuery['bool']>;
+type NonNullableFilter = NonNullable<NonNullableQueryBool['filter']>;
+
+export interface GetQuerySearchRequest extends SearchRequest {
+  query: NonNullableQuery & {
+    bool: NonNullableQueryBool & {
+      filter: ExtractArrayType<NonNullableFilter>;
+    };
+  };
+  sort: NonNullableSort;
+}
+
+export const getQuery = ({
+  indexNames,
+  limit,
+  from,
+  outcome,
+  startDate,
+  endDate,
+}: GetQuery): GetQuerySearchRequest => {
```
NIT: We don't have to be very strict for the `getQuery` function return type. It is okay to change the query, as long as it is a valid `SearchRequest` object, so it can be passed to the ES client for the search request. I think it's better to use `SearchRequest` type directly, so we have the flexibility to change the implementation without having to adapt the types as well.

kapral18 committed:
refactor: rename routes and constants for index results

- Renamed constants for index results routes in `constants.ts`.
- Updated test files to reflect the new route names.
- Renamed route handler files to match the new route names.
- Adjusted import statements and references to the renamed routes and constants.

kapral18 committed:
refactor: Add total hits count to index results and rename limit to size

- Introduced a new helper function `getHitsTotal` to extract the total hits count from the search response.
- Updated the `get_index_results` route to include the total hits count in the response.
- Renamed the query parameter `limit` to `size` for consistency with Elasticsearch terminology.
- Modified tests to reflect the changes in the response structure and query parameter name.
- Updated the schema to use `t.partial` for optional query parameters and renamed `limit` to `size`.

kapral18 committed:
refactor: Update API routes for ECS data quality dashboard

- Changed API route from `/internal/ecs_data_quality_dashboard/results/indices_latest/{pattern}` to `/internal/ecs_data_quality_dashboard/results_latest/{pattern}`.
- Updated corresponding constants and function calls to reflect the new API routes.

kapral18 committed:
refactor: handle case when hits.total is 0 in getHitsTotal helper function

semd commented on x-pack/plugins/ecs_data_quality_dashboard/server/helpers/get_authorized_index_names.ts:
Thanks for extracting this logic, it's much cleaner ❤️ 
Would it make sense to have a unit test for it?

kapral18 replied:
ideally yes but it's already covered by unit tests of consuming files twice.

kapral18 committed:
refactor: rename result schema types for consistency

- Renamed `GetResultsIndicesPatternQuery` to `GetIndexResultsQuery`
- Renamed `GetResultsIndicesPatternParams` to `GetIndexResultsParams`
- Renamed `GetResultsIndicesLatestParams` to `GetIndexResultsLatestParams`
- Renamed `PostResultBody` to `PostIndexResultBody`
- Updated route validation to use the new schema type names
- Changed `size` and `from` query parameters to use `StringToPositiveNumber` type

angorayc commented on x-pack/plugins/ecs_data_quality_dashboard/server/helpers/get_authorized_index_names.ts:
```diff
@@ -0,0 +1,56 @@
+import type { IScopedClusterClient } from '@kbn/core-elasticsearch-server';
+import { checkIndicesPrivileges } from '../routes/results/privileges';
+
+export const getAuthorizedIndexNames = async (
```
Can we have a dedicated test for this function please
